atlasdb-container-test-utils/src/main/java/com/palantir/atlasdb/containers/Containers.java

```diff
@@ -0,0 +1,143 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.containers;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.rules.ExternalResource;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Sets;
+import com.jayway.awaitility.Awaitility;
+import com.palantir.atlasdb.testing.DockerProxyRule;
+import com.palantir.docker.compose.DockerComposeRule;
+import com.palantir.docker.compose.configuration.DockerComposeFiles;
+import com.palantir.docker.compose.configuration.ProjectName;
+import com.palantir.docker.compose.logging.LogCollector;
+import com.palantir.docker.compose.logging.LogDirectory;
+
+public class Containers extends ExternalResource {
+    private static final ProjectName PROJECT_NAME = ProjectName.fromString("atlasdbcontainers");
+    @VisibleForTesting
+    static final DockerProxyRule DOCKER_PROXY_RULE = new DockerProxyRule(PROJECT_NAME, Container.class);
+
+    private static final Set<Container> containersToStart = new HashSet<>();
+    private static final Set<Container> containersStarted = new HashSet<>();
+    private static final LoadingCache<String, String> dockerComposeFilesToTemporaryCopies =
+            CacheBuilder.<String, String>newBuilder()
+                    .build(CacheLoader.from(Containers::getDockerComposeFile));
+
+    private static volatile DockerComposeRule dockerComposeRule;
+    private static volatile LogCollector currentLogCollector;
+    private static volatile boolean shutdownHookAdded;
+    private static volatile boolean dockerProxyRuleStarted;
+
+    private final String logDirectory;
+
+    public Containers(Class<?> classToSaveLogsFor) {
+        logDirectory = LogDirectory.circleAwareLogDirectory(Paths.get(
+                "atlasdbcontainers",
+                classToSaveLogsFor.getSimpleName()).toString());
+    }
+
+    public Containers with(Container container) {
+        synchronized (Containers.class) {
+            containersToStart.add(container);
+        }
+        return this;
+    }
+
+    @Override
+    protected void before() throws Throwable {
```

Comment:
This is pretty long and difficult to read - we could break it up into many smaller functions